nc2zarr/writer.py

```diff
@@ -173,12 +201,90 @@ def _append_dataset(self, ds: xr.Dataset):
+                        drop=True)
+                    ds_new_only.to_zarr(self._output_store,
+                                        append_dim=append_dim,
+                                        consolidated=self._output_consolidated)
+                elif mode in (AppendMode.append_all, AppendMode.forbid_overlap):
+                    ds.to_zarr(self._output_store,
+                               append_dim=append_dim,
+                               consolidated=self._output_consolidated)
+                else:
+                    # This should never happen, but best to handle it anyway.
+                    raise NotImplementedError(
+                        f"Unhandled append mode {mode}"
+                    )
             else:
                 LOGGER.warning('Appending disabled, dry run!')
 
+    def _append_with_insertions(self, ds):
+        append_dim = self._output_append_dim
+        for i in range(ds.dims[append_dim]):
+            dataslice = ds.isel({append_dim: slice(i, i+1)})
+            insert_index, update_mode = \
+                find_slice(self._output_store, dataslice[append_dim],
+                           dimension=append_dim)
+            if update_mode == "append":
+                # Currently only works for local filesystem stores
+                # TODO: make append_slice fsspec-aware
+                append_slice(self._output_store.root, dataslice,
+                             dimension=append_dim)
+            elif update_mode == "insert":
+                # Currently only works for local filesystem stores
+                # TODO: make update_slice fsspec-aware
+                update_slice(self._output_store.root, insert_index,
+                             dataslice, update_mode, dimension=append_dim)
+            elif update_mode == "replace":
+                if self._output_append_mode is AppendMode.replace:
+                    # Currently only works for local filesystem stores
+                    # TODO: make update_slice fsspec-aware
+                    update_slice(self._output_store.root, insert_index,
+                                 dataslice, update_mode, dimension=append_dim)
+                # If the append mode is not "replace", it must be "retain" --
+                # so we do nothing, which retains the existing slice and
+                # discards the new one.
+            else:
+                # The remaining possible value is "create" -- but we've
+                # already confirmed that we have an existing store,
+                # so "create" is a "can't happen".
+                raise NotImplementedError(
+                    f"Unhandled update mode {update_mode}!")
+
+    def _check_append_allowed(self, ds: xr.Dataset) -> None:
+        output_ds = xr.open_zarr(self._output_store)
+        if self._output_append_mode is not AppendMode.append_all:
+            if not self._is_append_dim_increasing(output_ds):
```

Comment:
I've renamed the method to `_is_append_dim_monotonic_increasing` to make the condition more explicit (commit 8e069b4).

As to efficiency: this `_check_append_allowed` is not called for every time slice, only once per call to `write_dataset`. So I think it's fine, especially since we're only reading a single, one-dimensional variable which should in most cases fit into a single chunk. I suggest we keep it as is for now, and consider optimization options if it turns out to be a problem in practice. OK?

Reply:
> this _check_append_allowed is not called for every time slice, only once per call to `write_dataset`

That's not entirely right. `write_dataset` _is_ called for every input, see impl. in `Converter`. In CCI we had ten thousands of inputs each being one time slice.

> OK?

OK!